### PR TITLE
WIP docs for deepfreeze whatsnew

### PR DESCRIPTION
**Pull request by kumaraditya303**
Just some draft text.

### CODE REVIEW

File: Doc/whatsnew/3.11.rst [hunk 1]
```diff
@@ -425,8 +425,15 @@ runtime. Other optimizations not under this project are listed in `Optimizations
...
 
-Static objects
-~~~~~~~~~~~~~~
+Static Code objects
+~~~~~~~~~~~~~~~~~~~
+
+.. sectionauthor:: Kumar Aditya
+
+Python uses code objects to represent executable code for the interpreter
+which is allocated on heap by unmarshalling the bytecode. In Python 3.11,
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.11.rst:434]: This first line is pretty loaded. I don't mean to insult, but as a non-expert in this area, I'm quite lost.

To clarify, do you mean that the previous startup is:
```
marshalled code object (contains bytecode) -----(unmarshall/deserialize)------> heap code object
```
and with the new method, static code objects skip the first part and the unmarshalling phase?

**Review by Fidget-Spinner — COMMENTED** [review]
Woops. Sorry, I missed this in my notifications.

Other than code objects, are there more? I remember strings are one, not sure if there are anything else.

File: Doc/whatsnew/3.11.rst (hunk 1 above)
Comment (kumaraditya303) [reply Doc/whatsnew/3.11.rst:434]: Yeah, so here is the entire process:

Python caches the bytecode in `__pycache__` directory to skip the reading the source code  everytime the script is executed, it is converted to a heap allocated code objects which is what the interpreter operates on and these code objects are created by unmarshalling process which reads the bytecode and creates code objects.

So without deepfreeze python execution (assuming bytecode cache exists): 

```console
Read bytecode cache -> unmarshal -> Heap alloc code object -> ceval.c
```

With deepfreeze:
```console
Statically allocated code objects (1) -> ceval.c
```

So deepfreeze skips the first there stages of execution to speed up the execution.

(1) - there is some book keeping to ensure proper deallocation of these code objects as cpython currently does not support immortal objects.

Comment (kumaraditya303) [reply Doc/whatsnew/3.11.rst:434]: @Fidget-Spinner I'll update the docs, I had created this PR just so I don't forget about it. Feel free to ask any thing more about deepfreeze.

Comment (Fidget-Spinner) [reply Doc/whatsnew/3.11.rst:434]: Ah got it. The bookkeeping about fake "immortal" objects is cool. Didn't know about that.

File: Doc/whatsnew/3.11.rst [hunk 2]
```diff
@@ -425,8 +425,15 @@ runtime. Other optimizations not under this project are listed in `Optimizations
...
+statically allocated code objects of the core modules are generated to speed up
+startup of interpreter by 10-15% as this process skips the unmarshalling phase.
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.11.rst:436]: OK, incorporating your explanation and summarising, WDYT about this?
PS. Linking to this for the `__pycache__` part would be good: [compiled-python-files](https://docs.python.org/3/tutorial/modules.html#compiled-python-files)
```suggestion
Python caches bytecode in the ``__pycache__`` directory to speed up module loading.
Python module execution looks like this:

``Read __pycache__ -> Unmarshal -> Heap allocated code object -> Evaluate``

In Python 3.11, the core modules essential for Python startup are "frozen".
This means that their code objects (and bytecode) are statically allocated
by the interpreter. This reduces the steps in module execution process to this:

``Statically allocated code object -> Evaluate``

Interpreter startup is now 10-15% in Python 3.11.
```

File: Doc/whatsnew/3.11.rst (hunk 1 above)
Comment (kumaraditya303) [reply Doc/whatsnew/3.11.rst:434]: ## Without deepfreeze:
```mermaid
flowchart LR;
    id1[Bytecode cache __pycache__] --> id2[Unmarshal];
    id2 --> id3[Heap allocate code object];
    id3 --> id4[Evaluate];
```
----

## With deepfreeze:
```mermaid
flowchart LR;
    id1[Statically allocated code objects] --> id2[Evaluate];
```
----

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (kumaraditya303) [reply Doc/whatsnew/3.11.rst:436]: Yes, if rst supported mermaid diagrams it would have been better.

**Comment by Fidget-Spinner** [comment]
Merged in 77701feba07ed673948bdfe8c64d8c5a44444540. PS, I tried adding you as ``.. sectionauthor::``, but for some reason the directive doesn't render anything at all in HTML, so I gave up. If you find some way to make it work, please tell me.

I added you as co-author to the commit. Thanks!

**Pull request closed without merging**